dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusAArch64.java

```diff
@@ -105,163 +143,302 @@ public Object visitInstructionRow(LitmusAArch64Parser.InstructionRowContext ctx)
     }
 
     @Override
-    public Object visitMov(LitmusAArch64Parser.MovContext ctx) {
+    public Object visitMov(MovContext ctx) {
         Register register = programBuilder.getOrNewRegister(mainThread, ctx.rD, archType);
         Expression expr = ctx.expr32() != null ? (Expression)ctx.expr32().accept(this) : (Expression)ctx.expr64().accept(this);
-        return programBuilder.addChild(mainThread, EventFactory.newLocal(register, expr));
+        return add(EventFactory.newLocal(register, expr));
     }
 
     @Override
-    public Object visitCmp(LitmusAArch64Parser.CmpContext ctx) {
+    public Object visitCmp(CmpContext ctx) {
         Register register = programBuilder.getOrNewRegister(mainThread, ctx.rD, archType);
         Expression expr = ctx.expr32() != null ? (Expression)ctx.expr32().accept(this) : (Expression)ctx.expr64().accept(this);
         lastCmpInstructionPerThread.put(mainThread, new CmpInstruction(register, expr));
         return null;
     }
 
     @Override
-    public Object visitArithmetic(LitmusAArch64Parser.ArithmeticContext ctx) {
-        Register rD = programBuilder.getOrNewRegister(mainThread, ctx.rD, archType);
-        Register r1 = programBuilder.getOrErrorRegister(mainThread, ctx.rV);
-        Expression expr = ctx.expr32() != null ? (Expression)ctx.expr32().accept(this) : (Expression)ctx.expr64().accept(this);
-        return programBuilder.addChild(mainThread, EventFactory.newLocal(rD, expressions.makeIntBinary(r1, ctx.arithmeticInstruction().op, expr)));
+    public Object visitArithmetic(ArithmeticContext ctx) {
+        final Register r64 = programBuilder.getOrNewRegister(mainThread, ctx.rD, archType);
+        final Register register = getOrNewRegister32(ctx.rD32, r64, false, false);
```

Comment:
Why are we passing a reg64 to a method that (I assume by the name) is creating a reg32?

Reply:
I also found this method very confusing. The method can actually generate all of 8/16/32/64 bit registers...

Reply:
A more suitable name would be `shrinkRegisterIfRequired`.  The registers carrying dependencies between instructions are supposed to be the 64-bit ones.  The byte-width of load instructions was initially linked to the result register, however.  Now that that has changed, I can replace this facility with `EventFactory.newLoadWithAccessTypeAndMo(Register,Expression,Type,String)` and use only 64-bit registers.

Reply:
On second thought, this would fit into its own PR.  There are a lot of places, where `accessType` is not properly maintained, e.g. the `Compilation` pass.

Reply:
I agree. You should ignore `AccessType` for now: almost all code assumes that the value expression of stores and the result register of loads implicitly define the access type.
Since we never needed to argue about sizes of accesses in the absence of MSA, the access type didn't really serve a purpose so far.
Fixing this is a whole different issue.